### PR TITLE
getcwd_syscall() updates

### PR DESCRIPTION
**Pull request by ve1nard**
## Description

Fixes # (issue)

The following changes include a small fix for checking null pointers, more elaborate comments, and new unit tests for `getcwd_syscall()`.

### Type of change

- [ ] More detailed comments for `getcwd_syscall()`
- [ ] Added a condition to error-out null pointers passed as `buf` argument
- [ ] New unit tests for `getcwd_syscall()`

## How Has This Been Tested?

To run the tests, we need to run cargo test --lib command inside the safeposix-rust directory.

All the tests are present under this directory: lind_project/src/safeposix-rust/src/tests/fs_tests.rs

- Test A - `lut_lind_fs_getcwd_valid_args()`
- Test B - `ut_lind_fs_getcwd_invalid_args()`

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to a pull request and/or merged in other modules (native-client, lind-glibc, lind-project)

### CODE REVIEW

**Comment by rennergade** [comment]
@Anway-Agte can you review this before Rupesh does?

File: src/tests/fs_tests.rs [hunk 1]
```diff
@@ -595,7 +594,7 @@ pub mod fs_tests {
     }
 
     #[test]
-    pub fn ut_lind_fs_fcntl_dup(){
+    pub fn ut_lind_fs_fcntl_dup() {
```
Comment (Anway-Agte) [inline src/tests/fs_tests.rs]: Are these changes intentional?

File: src/tests/fs_tests.rs [hunk 2]
```diff
@@ -1352,35 +1351,91 @@ pub mod fs_tests {
...
+
+        //`/newcwd2` is a valid directory that we will create to check
+        //if calling `getcwd_syscall()` with invalid arguments
+        //correctly returns the required errors.
+        //`newcwdsize` is the size of the string needed to store the new
+        //current woring directory `/newcwd1`
+        let newcwdsize = "/newcwd2\0".as_bytes().to_vec().len();
+        let newcwdsize_u32: u32 = newcwdsize as u32;
+        //Creating the new current working directory, checking if it exits
+        //and changing to it
```
Comment (Anway-Agte) [inline src/tests/fs_tests.rs:1466]: small typo - think it should be "Exists"

File: src/tests/fs_tests.rs [hunk 3]
```diff
@@ -1352,35 +1351,91 @@ pub mod fs_tests {
...
+        //Acquiring a lock on TESTMUTEX prevents other tests from running
+        //concurrently, and also performs clean env setup
+        let _thelock = setup::lock_and_init();
+        let cage = interface::cagetable_getref(1);
+
+        //`/newcwd2` is a valid directory that we will create to check
+        //if calling `getcwd_syscall()` with invalid arguments
+        //correctly returns the required errors.
+        //`newcwdsize` is the size of the string needed to store the new
+        //current woring directory `/newcwd1`
```
Comment (Anway-Agte) [inline src/tests/fs_tests.rs]: small typo - "working"

File: src/tests/fs_tests.rs [hunk 4]
```diff
@@ -1352,35 +1351,91 @@ pub mod fs_tests {
...
 
-        let mut buf = vec![0u8; needed];
+        //Checking if retrieving the current working directory
+        //`/newcwd1` by passing a string long enough to store
+        //the directory succeeds
+        //`newcwdsize` is the size of the string needed to store the new
+        //current woring directory `/newcwd1`
+        let newcwdsize = "/newcwd1\0".as_bytes().to_vec().len();
+        let newcwdsize_u32: u32 = newcwdsize as u32;
+        //Creating the new current working directory and checking if it exits
```
Comment (Anway-Agte) [inline src/tests/fs_tests.rs]: small typo - i think this should be "exists"

**Review by Anway-Agte — CHANGES_REQUESTED** [review]

File: src/tests/fs_tests.rs (hunk 1 above)
Comment (ve1nard) [reply src/tests/fs_tests.rs]: this is the result of running cargofmt

File: src/tests/fs_tests.rs [hunk 5]
```diff
@@ -1412,34 +1412,89 @@ pub mod fs_tests {
     #[test]
     pub fn ut_lind_fs_dir_chdir_getcwd() {
         //acquiring a lock on TESTMUTEX prevents other tests from running concurrently,
-        // and also performs clean env setup
+        //and also performs clean env setup
```
Comment (Anway-Agte) [inline src/tests/fs_tests.rs:1415]: It'd be good if you can explain in a couple of lines what happens in the clean env setup

File: src/tests/fs_tests.rs [hunk 6]
```diff
@@ -1412,34 +1412,89 @@ pub mod fs_tests {
...
         let bufptr: *mut u8 = &mut buf[0];
 
+        //First, we change to the root directory and call `getcwd_syscall()`
+        //to check if the root current working directory is successfully
+        //retrieved
         assert_eq!(cage.chdir_syscall("/"), 0);
-        assert_eq!(cage.getcwd_syscall(bufptr, 0), -(Errno::ERANGE as i32));
-        assert_eq!(cage.getcwd_syscall(bufptr, 1), -(Errno::ERANGE as i32));
         assert_eq!(cage.getcwd_syscall(bufptr, 2), 0);
         assert_eq!(std::str::from_utf8(&buf).unwrap(), "/\0\0\0\0\0\0\0\0");
```
Comment (Anway-Agte) [inline src/tests/fs_tests.rs:1439]: Why are there multiple null terminators in the string that represents the root directory?

File: src/tests/fs_tests.rs [hunk 7]
```diff
@@ -1412,34 +1412,89 @@ pub mod fs_tests {
...
+        //buf is not a null pointer` error
+        //Creating a string large enough to store the new current
+        //working directory and filling it with 0's
+        let mut buf = vec![0u8; newcwdsize];
+        let bufptr: *mut u8 = &mut buf[0];
+        assert_eq!(cage.getcwd_syscall(bufptr, 0), -(Errno::EINVAL as i32));
 
-        assert_eq!(cage.getcwd_syscall(bufptr, 0), -(Errno::ERANGE as i32));
+        //Checking if passing a valid string pointer and a non-zero size smaller
+        //than the size of the current working direcotry plus the terminating null
```
Comment (Anway-Agte) [inline src/tests/fs_tests.rs]: "directory"

File: src/tests/fs_tests.rs [hunk 8]
```diff
@@ -1412,34 +1412,89 @@ pub mod fs_tests {
...
+        );
+
+        //Checking if passing a valid string pointer and a size of 0 to
+        //`getcwd_syscall()` correctly results in `The size argument is zero and
+        //buf is not a null pointer` error
+        //Creating a string large enough to store the new current
+        //working directory and filling it with 0's
+        let mut buf = vec![0u8; newcwdsize];
+        let bufptr: *mut u8 = &mut buf[0];
+        assert_eq!(cage.getcwd_syscall(bufptr, 0), -(Errno::EINVAL as i32));
```
Comment (Anway-Agte) [inline src/tests/fs_tests.rs:1478]: why do we negate the sign of the error number?

File: src/tests/fs_tests.rs [hunk 9]
```diff
@@ -1412,34 +1412,89 @@ pub mod fs_tests {
...
+        //character to `getcwd_syscall()` correctly results in `The bufsize argument
+        //is less than the length of the absolute pathname of the working directory,
+        //including the terminating null byte` error
         assert_eq!(
-            cage.getcwd_syscall(bufptr, needed_u32 - 1),
+            cage.getcwd_syscall(bufptr, newcwdsize_u32 - 1),
             -(Errno::ERANGE as i32)
         );
-        assert_eq!(cage.getcwd_syscall(bufptr, needed_u32), 0);
-        assert_eq!(std::str::from_utf8(&buf).unwrap(), "/subdir1\0");
```
Comment (Anway-Agte) [inline src/tests/fs_tests.rs:1489]: In this function - do we also need to check if the getcwd syscall handles the case when the pointer passed to it is not a pointer to a string but a different variable?

**Review by Anway-Agte — APPROVED** [review]

File: src/tests/fs_tests.rs (hunk 6 above)
Comment (ve1nard) [reply src/tests/fs_tests.rs:1439]: the root directory is represented by "/". everything else is just to fill up the remaining space for the sake of the example

**Review by rupeshkoushik07 — APPROVED** [review]
looks good to me!

File: src/safeposix/syscalls/fs_calls.rs [hunk 10]
```diff
@@ -3599,20 +3599,78 @@ impl Cage {
...
-        let mut bytes: Vec<u8> = self.cwd.read().to_str().unwrap().as_bytes().to_vec();
-        bytes.push(0u8); //Adding a null terminator to the end of the string
-        let length = bytes.len();
-
-        if (bufsize as usize) < length {
-            return syscall_error(Errno::ERANGE, "getcwd", "the length (in bytes) of the absolute pathname of the current working directory exceeds the given size");
+        //The first two conditions help to error-out quickly if either
+        //the pointer to the string is a null pointer or the specified
+        //size of the string is 0.
+        if !buf.is_null() {
```
Comment (rennergade) [inline src/safeposix/syscalls/fs_calls.rs]: I see why you added this but its actually redundant. We do any checks for EFAULT in the type checker as well as within the actual runtime. I'd remove this check and note that its checked in types.rs.

**Review by rennergade — CHANGES_REQUESTED** [review]
Minor comment about the EFAULT check. I also don't really know why there are so many extra files in this PR.

Otherwise really great job on this one, incredibly thorough.

**Review by rennergade — APPROVED** [review]
Awesome. Great job!

**Pull request merged**